verification/spec/connection-handshake/Environment.tla

```diff
@@ -0,0 +1,233 @@
+    or bufChainB. *)
+InitEnv ==
+    /\ maliciousEnv = TRUE
+    /\ \/ /\ bufChainA \in {<<msg>> : msg \in InitMsgs(chmA!ConnectionEnds, chmB!ConnectionEnds)}
+          /\ bufChainB = <<>>   (* Empty buffer. *)
+       \/ /\ bufChainB \in {<<msg>> : msg \in InitMsgs(chmB!ConnectionEnds, chmA!ConnectionEnds)}
+          /\ bufChainA = <<>>
+
+
+(* The environment overwrites the buffer of one of the chains. 
+   This interferes with the CH protocol in two ways:
+    1. by introducing additional messages that are incorrect,
+    2. by dropping correct messages (overwritting them).
+    
+    Without the first constraint, on the "Len(bufChainA)" and "Len(bufChainB)",
+    Env could fill buffers (DoS attack). This can lead to a deadlock, because
+    chains will simply be unable to reply to each other.
+ *)
+MaliciousNextEnv ==
+    \/ /\ Len(bufChainA) < MaxBufLen - 1
+       /\ bufChainA' \in
+            {Append(bufChainA, arbitraryMsg) : 
+                arbitraryMsg \in ConnectionHandshakeMessages} 
+       /\ UNCHANGED bufChainB
+    \/ /\ Len(bufChainA) < MaxBufLen - 1
+       /\ bufChainB' \in
+            {Append(bufChainB, arbitraryMsg) :
+                arbitraryMsg \in ConnectionHandshakeMessages}
+       /\ UNCHANGED bufChainA
+
+
+NextEnv ==
+    \/ /\ maliciousEnv = TRUE
+       /\ MaliciousNextEnv
+       /\ UNCHANGED maliciousEnv
+    \/ /\ maliciousEnv = TRUE
+       /\ maliciousEnv' = FALSE
+       /\ UNCHANGED<<bufChainA, bufChainB>>
+
+
+CHDone ==
+    /\ storeChainA.connection.state = "INIT" (* TODO: should be OPEN *)
+    /\ storeChainB.connection.state = "INIT"
+    /\ UNCHANGED <<allVars>>
+
+(******************************************************************************
+ Main spec.
+ The system comprises the connection handshake module & environment.
+ *****************************************************************************)
+
+
+Init ==
+    /\ chmA!Init("chainA")
+    /\ chmB!Init("chainB")
+    /\ InitEnv
+
+
+\* The two CH modules and the environment alternate their steps.
+Next ==
+    \/ CHDone
+    \/ NextEnv /\ UNCHANGED <<chainStoreVars>>
+    \/ chmA!Next /\ UNCHANGED <<storeChainB, maliciousEnv>>
+    \/ chmB!Next /\ UNCHANGED <<storeChainA, maliciousEnv>>
+
+
+Spec ==
+    /\ Init
+    /\ [][Next]_<<allVars>>
+    /\ WF_chainAVars(chmA!Next)
+    /\ WF_chainBVars(chmB!Next)
+
+
+TypeInvariant ==
+    /\ bufChainA \in ConnectionHandshakeMessages
+    /\ bufChainB \in ConnectionHandshakeMessages
+
+
+\* Liveness property.
+Termination ==
+    <> ~ maliciousEnv
+        => <> /\ storeChainA.connection.state = "INIT" (* TODO: should be OPEN *)
+              /\ storeChainB.connection.state = "INIT"
+
+
+\* Safety property.
+\* If the connections in the two chains are not null, then the
+\* connection parameters must always match.
+ConsistencyInv ==
+    \/ storeChainA.connection = chmA!nullConnection
+    \/ storeChainB.connection = chmB!nullConnection
+    \/ storeChainA.connection.parameters 
```

Comment:
I guess we only want connection ends to match if connection is in OPEN state at both sides. In that case connection is open and established and it should satisfy this property. In theory we don't care if at some step they are not consistent: imagine a case where connections ends are simultaneously initialled with different ids. This is not problematic if later they converge to a consists ids.

Reply:
There is currently no negotiation allowed in the IBC spec on IDs for connections and channels. If you make a mistake and mismatch the IDs, e.g. you send `openInit` on A and B without matching, they will be successful but will stay stuck in `INIT` state and another `openInit` will not fix this. The `connectionEnd` state stays in the store forever, they cannot be changed or removed.

Reply:
>  In theory we don't care if at some step they are not consistent: imagine a case where connections ends are simultaneously initialled with different ids. This is not problematic if later they converge to a consists ids.

I had an earlier version that was allowing "temporary inconsistency". But I ran into the problem that the connection ends would never reconcile in fact. To allow them to reconcile, we would need to modify our spec to capture multiple connections at the same time.

Reply:
We are talking here about property it should ensure, so I am not saying protocol would be able to make it for a mistake (inconsistency). We care if this holds only if it's init at both side as at that point it becomes active (?), similar like we care about consensus only after we decide.